### PR TITLE
Handling the notification of events by the Orion Context Broker

### PR DESCRIPTION
**Pull request by gtorodelvalle**
- Implements https://github.com/telefonicaid/IoT-STH/issues/18
- 100% tests passed
- Assigned to @frbattid 
- Feedback also requested from @iariasleon

### CODE REVIEW

File: src/sth_server.js [hunk 1]
```diff
@@ -141,6 +155,86 @@
...
+          if (request.payload && request.payload.contextResponses &&
+            Array.isArray(request.payload.contextResponses)) {
+            contextResponses = request.payload.contextResponses;
+            for (var i = 0; i < contextResponses.length; i++) {
+              if (contextResponses[i].contextElement &&
+                contextResponses[i].contextElement.attributes &&
+                Array.isArray(contextResponses[i].contextElement.attributes)) {
+                attributes = contextResponses[i].contextElement.attributes;
+                for (var j = 0; j < attributes.length; j++) {
+                  if (isNaN(attributes[j].value)) {
```
Comment (frbattid) [inline src/sth_server.js]: Any idea about how much time consuming is this operation?

File: src/sth_database.js [hunk 2]
```diff
@@ -236,6 +236,227 @@
...
+   * @param {string} entityId The entity id associated to updated attribute
+   * @param {string} entityType The entity type associated to the updated attribute
+   * @param {string} attributeId The updated attribute id
+   * @param {string} attributeType The updated attribute type
+   * @param {string} attributeValue The updated attribute value
+   * @param {Function} callback Function to call once the operation completes
+   */
+  function storeRawData(
+    timestamp, entityId, entityType, attributeId, attributeType, attributeValue, callback) {
+    var collectionName4Events = getCollectionName4Events(entityId, attributeId);
```
Comment (frbattid) [inline src/sth_database.js]: This seems to be creating a collection name based on the entityId and the attributeId... this should be aligned (later or now, I don't know) with the naming conventions we are discussing offline.

File: src/sth_database.js [hunk 3]
```diff
@@ -236,6 +236,227 @@
...
+   * @param {string} entityId The entity id associated to updated attribute
+   * @param {string} entityType The entity type associated to the updated attribute
+   * @param {string} attributeId The updated attribute id
+   * @param {string} attributeType The updated attribute type
+   * @param {string} attributeValue The updated attribute value
+   * @param {Function} callback Function to call once the operation completes
+   */
+  function storeAggregatedData(
+    timestamp, entityId, entityType, attributeId, attributeType, attributeValue, callback) {
+    var collectionName4Aggregated = getCollectionName4Aggregated(entityId, attributeId);
```
Comment (frbattid) [inline src/sth_database.js]: This seems to be creating a collection name based on the entityId and the attributeId... this should be aligned (later or now, I don't know) with the naming conventions we are discussing offline.

File: README.md [hunk 4]
```diff
@@ -116,6 +120,75 @@ This information has as its origin the 46nd minute, of the 2nd hour of February,
...
+        {
+            "type": "Room",
+            "isPattern": "false",
+            "id": "Room1-gtv"
+        }
+    ],
+    "attributes": [
+        "temperature"
+    ],
+    "reference": "http://sth.host:1028/STH/v1/notify",
```
Comment (iariasleon) [inline README.md]: I recomend:
-  change :1028  to :sth.port 
- And in Cygnus we use the path only /notify, if we want that notifications STH are similar to notifications Cygnus.

File: README.md [hunk 5]
```diff
@@ -116,6 +120,75 @@ This information has as its origin the 46nd minute, of the 2nd hour of February,
...
+* <a href="https://github.com/telefonicaid/fiware-cygnus/tree/master/flume#orion-subscription" href="_blank">Orion subscription</a>
+
+The another option to update the aggregated time series information consists on directly subscribing the STH component
+to the Orion Context Broker to receive the corresponding notifications and generate and update the aggregated data.
+
+To subscribe the STH component to the Orion Context Broker as a way to receive attribute value updates, the following
+curl command can be used:
+
+<pre>
+curl orion.contextBroker.host:1026/v1/subscribeContext -s -S --header 'Content-Type: application/json' --header 'Accept: application/json' -d @- &lt;&lt;EOF
```
Comment (iariasleon) [inline README.md]: Also can to use Fiware-Service and Fiware-ServicePath headers

File: README.md (hunk 4 above)
Comment (gtorodelvalle) [inline README.md]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/58fa15d352bf6c8aac640ec5d680c2ca92115060

File: README.md (hunk 5 above)
Comment (gtorodelvalle) [inline README.md]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/58fa15d352bf6c8aac640ec5d680c2ca92115060

File: README.md [hunk 6]
```diff
@@ -164,6 +242,22 @@ without filtering out the empty results, use:
 
 <pre> PORT=7777 DB_HOST=mymongo.com DB_PORT=27777 FILTER_OUT_EMPTY=false npm start</pre>
 
+Of special interest is the DATA_MODEL environment variable. Currently, the STH component supports 3 possible data distribution
```
Comment (frbattid) [inline README.md]: I really like this section!

NTC

File: src/sth_database.js [hunk 7]
```diff
@@ -5,7 +5,109 @@
...
+          timestamp: Date,
+          attributeName: String,
+          attributeType: String,
+          attributeValue: Number
+        });
+        break;
+      case sthConfig.DATA_MODELS.COLLECTIONS_PER_ATTRIBUTE:
+        eventSchema = mongoose.Schema({
+          timestamp: Date,
+          attributeType: String,
```
Comment (frbattid) [inline src/sth_database.js]: Why is this added to the collection? When DATA_MODEL=COLLECTION-PER-ATTRIBUTE this kind of information seems to be added to the name of the collection:

```
case sthConfig.DATA_MODELS.COLLECTIONS_PER_ATTRIBUTE:
   return sthConfig.COLLECTION_PREFIX + '_' + servicePath + '_' + entityId + (entityType ? '_' + entityType : '') +
      '_' + attributeId + (attributeType ? '_' + attributeType : '');
```

I mean, regarding other pieces of information (entotyId, entityType, attrName), it seems the rule is: "if appearing in the collection name, then it is not added in the document content".

Comment (gtorodelvalle) [inline src/sth_database.js]: Really nice catch! ;) Yeah, that's a reminiscence from the past.. ;) Thank you very much for noticing!

Fixed in https://github.com/telefonicaid/IoT-STH/commit/afdb5eb5d0b51a6c0f4d4ee40b09777bc866fd3b

Comment (gtorodelvalle) [inline src/sth_database.js]: The attributeType field has been brought back to life :p as a consequence of the discussion we've had with @fgalan where he told us that the attribute type of a concrete attribute name could change over time and consequently cannot be used to generate the name of the collections but stored in the docs themselves ;)

Comment (fgalan) [inline src/sth_database.js]: Yes, that's correct: the attribute type of a given attribute of a given entity could be changed at Orion using its API.

File: README.md [hunk 8]
```diff
@@ -152,6 +222,15 @@ The script accepts the following parameters as environment variables:
 - LOG_TO_FILE: A flag indicating if the logs should be sent to a file. Optional. Default value: true.
 - LOG_DIR: The path to a directory where the log file will be searched for or created if it does not exist. Optional. Default value: "./log".
 - LOG_FILE_NAME: The name of the file where the logs will be stored. Optional. Default value: "sth_app.log".
+- SERVICE_PREFIX: The prefix to be added to the service for the creation of the databases. More information below. Optional. Default value: "sth".
```
Comment (frbattid) [inline README.md]: I thinks it is better "DB_PREFIX", since we are going to prefix the database name, independently of how it is created.

File: src/sth_database.js [hunk 9]
```diff
@@ -8,53 +8,137 @@
...
+   * Declares the Mongoose schemas.
+   */
+  function defineSchemas() {
+    switch (sthConfig.DATA_MODEL) {
+      case sthConfig.DATA_MODELS.COLLECTIONS_PER_SERVICE_PATH:
+        eventSchema = mongoose.Schema({
+          timestamp: Date,
+          entityId: String,
+          entityType: String,
+          attributeName: String,
```
Comment (frbattid) [inline src/sth_database.js]: We agreed to call this: attrName, attrType and attrValue.

File: src/sth_database.js [hunk 10]
```diff
@@ -8,53 +8,137 @@
...
+          attributeType: String,
+          attributeValue: Number
+        });
+        break;
+      case sthConfig.DATA_MODELS.COLLECTIONS_PER_ENTITY:
+        eventSchema = mongoose.Schema({
+          timestamp: Date,
+          attributeName: String,
```
Comment (frbattid) [inline src/sth_database.js]: We agreed to call this: attrName, attrType and attrValue.

File: src/sth_database.js (hunk 7 above)
Comment (frbattid) [inline src/sth_database.js]: attrType instead

File: src/sth_database.js [hunk 11]
```diff
@@ -8,53 +8,137 @@
...
+    }
+
+    switch (sthConfig.DATA_MODEL) {
+      case sthConfig.DATA_MODELS.COLLECTIONS_PER_SERVICE_PATH:
+        aggregatedSchema = mongoose.Schema({
+          _id: {
+            type: {
+              entityId: String,
+              entityType: String,
+              attributeName: String,
```
Comment (frbattid) [inline src/sth_database.js]: attrName instead

File: src/sth_database.js [hunk 12]
```diff
@@ -8,53 +8,137 @@
...
+            min: Number,
+            max: Number
+          }]
+        });
+        break;
+      case sthConfig.DATA_MODELS.COLLECTIONS_PER_ENTITY:
+        aggregatedSchema = mongoose.Schema({
+          _id: {
+            type: {
+              attributeName: String,
```
Comment (frbattid) [inline src/sth_database.js]: attrName instead

File: README.md (hunk 8 above)
Comment (gtorodelvalle) [inline README.md]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/4a37bf6666339a8e7d137ba3cb62fca558080ab0 Thanks @frbattid ! ;)

File: src/sth_database.js (hunk 9 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/4a37bf6666339a8e7d137ba3cb62fca558080ab0 Thanks @frbattid ! ;)

File: src/sth_database.js (hunk 7 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/4a37bf6666339a8e7d137ba3cb62fca558080ab0 Thanks @frbattid ! ;)

File: src/sth_database.js (hunk 10 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/4a37bf6666339a8e7d137ba3cb62fca558080ab0 Thanks @frbattid ! ;)

File: src/sth_database.js (hunk 11 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/4a37bf6666339a8e7d137ba3cb62fca558080ab0 Thanks @frbattid ! ;)

File: src/sth_database.js (hunk 12 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: Fixed in https://github.com/telefonicaid/IoT-STH/commit/4a37bf6666339a8e7d137ba3cb62fca558080ab0 Thanks @frbattid ! ;)

File: src/sth_database.js (hunk 9 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: In fact, fixed in https://github.com/telefonicaid/IoT-STH/commit/d45f39bd4cd9d62980a1d3bef5cec0adb5fe09c6

File: src/sth_database.js (hunk 10 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: In fact, fixed in https://github.com/telefonicaid/IoT-STH/commit/d45f39bd4cd9d62980a1d3bef5cec0adb5fe09c6

File: src/sth_database.js (hunk 11 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: In fact, fixed in https://github.com/telefonicaid/IoT-STH/commit/d45f39bd4cd9d62980a1d3bef5cec0adb5fe09c6

File: src/sth_database.js (hunk 12 above)
Comment (gtorodelvalle) [inline src/sth_database.js]: In fact, fixed in https://github.com/telefonicaid/IoT-STH/commit/d45f39bd4cd9d62980a1d3bef5cec0adb5fe09c6

**Comment by frbattid** [comment]
LGTM
